parser_api/example.rb

```diff
@@ -0,0 +1,173 @@
+      end
+    end
+  end
+
+  class MyFactory < Puppet::Pops::Validation::Factory
+    # Produces the checker to use
+    def checker(diagnostic_producer)
+      MyChecker.new(diagnostic_producer)
+    end
+
+    # Produces the label provider to use.
+    #
+    def label_provider
+      # We are dealing with AST, so the existing one will do fine.
+      # This is what translates objects into a meaningful description of what that thing is
+      #
+      Puppet::Pops::Model::ModelLabelProvider.new()
+    end
+
+    # Produces the severity producer to use. Here it is configured what severity issues have
+    # if they are not all errors. (If they are all errors this method is not needed at all).
+    #
+    def severity_producer
+      # Gets a default severity producer that is then configured below
+      p = super
+
+      # Configure each issue that should **not** be an error
+      #
+      p[Issues::INVALID_WORD]                 = :warning
+
+      # examples of what may be done here
+      # p[Issues::SOME_ISSUE]           = <some condition> ? :ignore : :warning
+      # p[Issues::A_DEPRECATION]        = :deprecation
+
+      # return the configured producer
+      p
+    end
+
+    # Allow simpler call when not caring about getting the actual acceptor
+    def diagnostic_producer(acceptor=nil)
+      acceptor.nil? ? super(Puppet::Pops::Validation::Acceptor.new) : super(acceptor)
+    end
+  end
+
+  # We create a diagnostic formatter that outputs the error with a simple predefined
+  # format for location, severity, and the message. This format is a typical output from
+  # something like a linter or compiler.
+  # (We do this because there is a bug in the DiagnosticFormatter's `format` method prior to
+  # Puppet 4.9.0. It could otherwise have been used directly.
+  #
+  class Formatter < Puppet::Pops::Validation::DiagnosticFormatter
+    def format(diagnostic)
+      "#{format_location(diagnostic)} #{format_severity(diagnostic)}#{format_message(diagnostic)}"
+    end
+  end
+end
+
+# -- Example usage of the new validator
+
+# Get a parser
+parser = Puppet::Pops::Parser::EvaluatingParser.singleton
+
+# parse without validation
+result = parser.parser.parse_string('$x = if 1 < 2 { smaller } else { bigly }', 'testing.pp')
+result = result.model
+
+# validate using the default validator and get hold of the acceptor containing the result
+acceptor = parser.validate(result)
+
+# -- At this point, we have done everything `puppet parser validate` does except report the errors
+# and raise an exception if there were errors.
+
+# The acceptor may now contain errors and warnings as found by the standard puppet validation.
+# We could look at the amount of errors/warnings produced and decide it is too much already
+# or we could simply continue. Here, some feedback is printed:
+#
+puts "Standard validation errors found: #{acceptor.error_count}"
+puts "Standard validation warnings found: #{acceptor.warning_count}"
+
```

Comment:
I would start with this example since it shows how to do a complete parse/validate using what's "in the box". Once that's covered, it's easier to understand that the objective with the `MyValidation` module is to go beyond that. It's also easier to understand how things fit together.